### PR TITLE
Add visual density to the gallery options

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This PR adds an option to the gallery options that sets the visual density to one of "standard" (the default), "comfortable", or "compact".

## Tests

- Added a test to exercise the new option.

## Breaking Change

- [X] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/test/material/raw_material_button_test.dart [hunk 1]
```diff
@@ -525,7 +525,7 @@ void main() {
                 visualDensity: visualDensity,
                 key: key,
                 onPressed: () {},
-                child: useText ? const Text('Text', key: childKey) : Container(key: childKey, width: 100, height: 100, color: const Color(0xffff0000)),
+                child: useText ? const Text('Text') : Container(width: 100, height: 100, color: const Color(0xffff0000)),
```
Comment (shihaohong) [inline packages/flutter/test/material/raw_material_button_test.dart]: It doesn't seem like this change was intentional

File: examples/flutter_gallery/test/drawer_test.dart [hunk 2]
```diff
@@ -54,16 +54,34 @@ void main() {
     // Switch back to system theme setting: first menu button, choose 'System Default'
     await tester.tap(find.byIcon(Icons.arrow_drop_down).first);
     await tester.pumpAndSettle();
-    await tester.tap(find.text('System Default').at(1));
+    await tester.tap(find.text('System Default').at(2));
```
Comment (shihaohong) [inline examples/flutter_gallery/test/drawer_test.dart]: Is there a way to particularly target this instance of `Text('System Default')` here and elsewhere? I was thinking something like: 
```
  find.descendant(
    of: find.byType(PopupMenu),
    matching: find.text('System Default')
  )
```
However, it seems both _PopupMenu and the _VisualDensityItem widgets are private, so I'm not sure if it's possible

**Comment by gspencergoog** [comment]
I think this is ready for another look.

**Review by shihaohong — APPROVED** [review]
LGTM!

File: examples/flutter_gallery/test/drawer_test.dart (hunk 2 above)
Comment (shihaohong) [reply examples/flutter_gallery/test/drawer_test.dart]: That's pretty neat, I'll definitely keep this in mind when I need it in the future!

**Comment by flar** [comment]
This PR failed the analyzer on linux:

```
▌04:44:01▐ RUNNING: cd .; bin/flutter analyze --dartdocs --flutter-repo
Analyzing 3 directories...                                      
  error • A value of type 'dynamic' can't be assigned to a variable of type 'GalleryVisualDensityValue' • examples/flutter_gallery/lib/gallery/scales.dart:50:50 • invalid_assignment
  error • A value of type 'Widget' can't be assigned to a variable of type 'MaterialApp' • examples/flutter_gallery/test/drawer_test.dart:73:11 • invalid_assignment
  error • A value of type 'Widget' can't be assigned to a variable of type 'MaterialApp' • examples/flutter_gallery/test/drawer_test.dart:83:11 • invalid_assignment
3 issues found. (ran in 278.8s)
▌04:48:43▐ ELAPSED TIME: 4min 41.944s for bin/flutter analyze --dartdocs --flutter-repo in .
```

**Comment by franciscojma86** [comment]
@flar why wasn't this reverted in the end? Just curious since seems like it might have introduced some performance regressions so not sure if it might be related.

**Comment by flar** [comment]
The build failures were fixed by someone else before I merged: https://github.com/flutter/flutter/pull/47594

**Comment by franciscojma86** [comment]
I see, thanks!

**Comment by flar** [comment]
> @flar why wasn't this reverted in the end? Just curious since seems like it might have introduced some performance regressions so not sure if it might be related.

I'm just seeing some performance regressions from around this time so you may be right, but they are subtle enough and the benchmarks themselves are noisy enough that it is hard to point at a specific commit that could have been the cause.

**Pull request merged**